### PR TITLE
Add AnnotationExcluder to be able to reuse exluding by annotation mechanism.

### PR DESCRIPTION
**Pull request by vanniktech**
Should ease implementing #377

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    private val excludes: Excludes
+) {
+  private var resolvedAnnotations = root.importList
+      ?.imports
+      ?.filterNot { it.isAllUnder }
+      ?.mapNotNull { it.importedFqName?.asString() }
+      ?.map { Pair(it.split(".").last(), it) }
+      ?.toMap()
+
+  fun shouldExclude(annotations: List<KtAnnotationEntry>) =
```
Comment (arturbosch) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Please add a test case for this feature :)

Comment (vanniktech) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Should I just copy the ones from LateInit then? I thought that relying on that for the moment is enough.

Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Extract it? Its much clearer having a test case for this too, don't you think?

Comment (vanniktech) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Sure, will do when I get time to this

Comment (vanniktech) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: How can I get a `KtAnnotationEntry` in tests?

Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Try the factory: `FACTORY.createAnnotationEntry("")`, sorry for late response!

**Comment by vanniktech** [comment]
Added tests. However I need a way to compile a String to a KtFile. I tried adding the test dependency however it's failing with:

> e: Supertypes of the following classes cannot be resolved. Please make sure you have the required dependencies in the classpath:
    class io.gitlab.arturbosch.detekt.test.KtTestCompiler, unresolved supertypes: io.gitlab.arturbosch.detekt.core.KtCompiler

Anyone has a hint?

File: detekt-test/src/main/kotlin/io/gitlab/arturbosch/detekt/test/KtTestCompiler.kt [hunk 2]
```diff
@@ -11,7 +11,7 @@ import java.nio.file.Paths
  *
  * @author Artur Bosch
  */
-internal object KtTestCompiler : KtCompiler(Paths.get(resource("/"))) {
```
Comment (arturbosch) [inline detekt-test/src/main/kotlin/io/gitlab/arturbosch/detekt/test/KtTestCompiler.kt]: There is a global function `compileContentForTest` for this. This class may change so `internal` should stay.

**Comment by vanniktech** [comment]
Now that I can run the tests, I'm seeing that the `createAnnotationEntry` fails since the modifierList is internally null:

```kotlin
fun createAnnotationEntry(text: String): KtAnnotationEntry {
  val modifierList = createProperty(text + " val x").modifierList
  return modifierList!!.annotationEntries.first()
}
```

**Comment by arturbosch** [comment]
Why don't you just parse some code as string and search for an AnnotationEntry?

**Comment by arturbosch** [comment]
Ok, just checked out your branch. You have to change your code to `FACTORY.createAnnotationEntry("@JvmField")` and it will work parsing.

**Comment by vanniktech** [comment]
That makes sense although it could fail a bit earlier if the string that's passed in is missing the `@`. In any case, thank you very much for the fix and I updated the PR.

**Comment by arturbosch** [comment]
Seems like the last test case is broken.

**Comment by vanniktech** [comment]
Yeah I just saw it. Will fix it.

**Comment by vanniktech** [comment]
@arturbosch this is now ready. Sorry it took that long :/

**Comment by arturbosch** [comment]
Thanks for the changes :)

**Pull request merged**